refactor: Gutenberg project practices alignment

dcalhoun opened the pull request:
## Description

> [!Tip]
> Given the quantity, there is likely little benefit in reviewing the file changes in its entirety. Reviewing individual commits may be more helpful. The list below provides a high-level description of the changes. 

Align with [Gutenberg](https://github.com/WordPress/gutenberg) project practices and improve overall consistency. These changes mirror general Gutenberg [package](https://github.com/WordPress/gutenberg/tree/2027fdbc060b24e59927dee5d09f95eb99a2fd76/packages) practices, but also the structure of existing mobile editor packages: [`react-native-editor`](https://github.com/WordPress/gutenberg/tree/2027fdbc060b24e59927dee5d09f95eb99a2fd76/packages/react-native-editor), [`react-native-bridge`](https://github.com/WordPress/gutenberg/tree/2027fdbc060b24e59927dee5d09f95eb99a2fd76/packages/react-native-bridge), and [`react-native-aztec`](https://github.com/WordPress/gutenberg/tree/2027fdbc060b24e59927dee5d09f95eb99a2fd76/packages/react-native-aztec). This is proposed for improving consistency and familiarity for Gutenberg contributors. 

- Consolidate `.gitignore` configuration
- Lowercase filenames
- Hoist `hook` to `src` directory
- Rename `misc` to `utils`
- Reorganize "helpers" to be less generic
- Remove file ending in import paths
- Hoist `lock-unlock` to root `src` directory
- Hoist JavaScript source to root directory
- Separate iOS files from JavaScript and Android files within `ios/`
- Separate Android files from JavaScript and iOS files within `android/`
- Improve Makefile casing consistency
- Relocate `app` to `components` directory 
- Relocate entry files to `src` directory
- Update README to include Android references 

## Testing Instructions

N/A, no user-facing changes.

dcalhoun committed:
refactor: Consolidate `.gitignore` configuration

Align with Gutenberg practices of consolidating Git configuration.

dcalhoun committed:
refactor: Lowercase filenames

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Hoist `hook` to `src` directory

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Rename `misc` to `utils`

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Reorganize "helpers" to be less generic

Align with Gutenberg project practices. Avoid generic filenames that
fail to describe intent or focus.

dcalhoun committed:
refactor: Remove file ending in import paths

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Hoist `lock-unlock` to root `src` directory

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Hoist JavaScript source to root directory

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Separate iOS files from JavaScript and Android files

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Separate Android files from JavaScript and iOS files

Align with Gutenberg project practices.

dcalhoun commented on src/utils/bridge.js:
```diff
@@ -70,24 +62,52 @@ export function openMediaLibrary(config) {
 	}
 }
 
-// FIXME: this was an attempt to fix an existing issue in Gutenberg , but it does it only
-// https://a8c.slack.com/archives/D0740HYKLUX/p1719841410651649
-//
-// export function removeHoverEffects() {
-//     for (let i = 0; i < document.styleSheets.length; i++) {
-//         let styleSheet = document.styleSheets[i];
-//         try {
-//             if (styleSheet.cssRules) {
-//                 for (let j = 0; j < styleSheet.cssRules.length; j++) {
-//                     let rule = styleSheet.cssRules[j];
-//                     if (rule.selectorText && rule.selectorText.includes(':hover')) {
-//                         styleSheet.deleteRule(j);
-//                         j--; // Adjust the index after deletion
-//                     }
-//                 }
-//             }
-//         } catch (e) {
-//             console.warn('Could not access stylesheet:', styleSheet, e);
-//         }
-//     }
-// }
```
Excluding documentation, this commented code removal is the sole code modification, all other changes relate to renaming and relocating existing code. I removed this particular code solely because its presence felt pointedly odd after renaming this particular file, and hope to identify a suitable replacement in core Gutenberg instead.

dcalhoun committed:
refactor: Improve Makefile casing consistency

dcalhoun committed:
refactor: Relocate `app` to `components` directory

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Rename `main` to `index`

Align with Gutenberg project practices.

dcalhoun committed:
refactor: Place entry files within `src` directory

Align with Gutenberg project practices.

dcalhoun committed:
docs: Update README

dcalhoun commented:
Realized I broke the WP-iOS integration by relocating the `Package.swift` file. I'll mark this as a draft until I address this. Sorry for the noise.

kean commented:
> [Package.swift → ios/Package.swift](https://github.com/wordpress-mobile/GutenbergKit/pull/37#diff-c796ca6d6e500b34c1c0117071fbbee2bedf2015e3c6175afba31be95fc6f740)

I'm pretty sure it needs to be at the root dir.

dcalhoun committed:
fix: Reinstate top-level Package.swift file

This must be in the root directory for Xcode to correctly source the
files. Additional configuration is also required to provide the new path
for the `ios` subdirectory.

dcalhoun committed:
fix: Remove now unnecessary directory changes in iOS build command

The package configuration file was relocated to the root directory.

dcalhoun committed:
fix: Repair GutenbergKit path for demo app

Path changed after relocating package configuration to root directory.

dcalhoun commented:
@kean thank you for your patience. This should now be ready for review. 🙇🏻‍♂️
